presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java

```diff
@@ -227,6 +180,53 @@ public boolean isDecomposable(FunctionRegistry functionRegistry)
                 .allMatch(InternalAggregationFunction::isDecomposable);
     }
 
+    public enum Step
+    {
+        PARTIAL(true, true),
+        FINAL(false, false),
+        INTERMEDIATE(false, true),
+        SINGLE(true, false);
+
+        private final boolean inputRaw;
+        private final boolean outputPartial;
+
+        Step(boolean inputRaw, boolean outputPartial)
+        {
+            this.inputRaw = inputRaw;
+            this.outputPartial = outputPartial;
+        }
+
+        public boolean isInputRaw()
+        {
+            return inputRaw;
+        }
+
+        public boolean isOutputPartial()
+        {
+            return outputPartial;
+        }
+
+        public static Step partialOutput(Step step)
+        {
+            if (step.isInputRaw()) {
+                return Step.PARTIAL;
+            }
+            else {
+                return Step.INTERMEDIATE;
+            }
+        }
+
+        public static Step partialInput(Step step)
+        {
+            if (step.isOutputPartial()) {
+                return Step.INTERMEDIATE;
+            }
+            else {
+                return Step.FINAL;
+            }
+        }
+    }
+
     public static class Aggregation
```

Comment:
shouldn't public static class be on top? That would mean that `Step` should also stay on top